uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/Test/Fabricator/Document.php

```diff
@@ -44,11 +43,7 @@ public static function fabricateWithAPI($params = []) {
    * @return array
    */
   private static function getDefaultParameters() {
-    $contactID = CRM_Core_Session::getLoggedInContactID();
-
-    if (!$contactID) {
-      $contactID = ContactFabricator::fabricate()['id'];
-    }
+    $contactID = self::getTestContactID();

```

Comment:
Was this change really necessary to make the tests pass or was it an attempt to improve things and remove the usage of the Contact Fabricator from HRCore?

I think that the previous version (where a contact would be created everytime) was bad, but I feel like this one is worse, because most of the time it will probably result in an unnecessary API call to check if a contact exists in the database.

I don't think it's a good idea to set default values for fields that are FKs to other entities. For these, it's better to make it explicit which are the entities that will be linked together.

So the options here are:
- Revert the previous version, but probably replace the HRCore fabricator with the T&A that I've suggested on another comment (We can then improve this later and remove the contact fabrication altogether)
- Change the tests so that we don't set a default value for contact ID fields (source_contact_id, target_contact_id and assignee_contact_id)

Please do whichever one that is faster to implement

Reply:
Yep, you asked me not to use CiviHR code here so I removed other references to the `ContactFabricator`. 

> I think that the previous version (where a contact would be created everytime) was bad, but I feel like this one is worse, because most of the time it will probably result in an unnecessary API call to check if a contact exists in the database.

I'd argue that a query call to check for a single contact is better than multiple queries and insertions related to the creation of a new contact.

> I don't think it's a good idea to set default values for fields that are FKs to other entities. For these, it's better to make it explicit which are the entities that will be linked together.

I'm conflicted on this myself, I really like the ability to just call `fabricate` and leave it to the defaults to decide what to use, but allowing for overriding of any of the defaults. For example for my tests for entity count in CiviHR I really don't care about any of the data, just that an entity is created.

My feeling is that creating a contact inside a document fabricator call is definitely wrong, and going down this path (of auto-creating required FK entities) will lead us into trouble later.

I can also see how using a random contact ID from the database or `user_contact_id` could lead to unexpected results, it's a matter of convenience (providing all defaults) vs clarity (requiring explicit FK IDs in `$params`) so probably better to go for the clarity option.

Reply:
If we're just comparing a single insert with a single select, then you're right and the performance of the select will probably be faster. Unfortunately, with your change, there will be one insert and one select (The insert has to happen before the call to the `fabricate()`, to create the contact that will be returned by the `get` call). Before this, there was only the insert call. Of course this all depends on how the fabricator will be used in the end. If the same contact is used for all the documents, then your version will probably work better. If not, then the previous would be better.

I really would prefer to not have all this details and performance differences hidden inside the fabricator. Not setting default values for these IDs would allows us to do things like this:

```php
// If I want different contacts for each document
$contact1 = Contact::fabricate();
$document1 = Document::fabricate([
  //other fields
  contact_id => $contact['id'],
  assigned_id => $contact['id'],
 //...
]);

$contact2 = Contact::fabricate();
$document2 = Document::fabricate([
  //other fields
  contact_id => $contact2['id'],
  assigned_id => $contact2['id'],
 //...
]);

// If I want the same contact for multiple documents
$singleContact = Contact::fabricate();
$contactParams = [
  contact_id => $contact2['id'],
  assigned_id => $contact2['id'],
];
$document1 = Document::fabricate(array_merge($contactParams, [
//document specific fields
]));

$document2 = Document::fabricate(array_merge($contactParams, [
//document specific fields
]));

// OR, if you really don't care about the contact
// 1 - disable FK check
// 2 - don't create any contact
$document = Document::fabricate([
  contact_id => 1,
  assignedd_id => 1,
  ...
]);
// enable FK check
```

The last example is used in many places in L&A to avoid us having to create entities that are really not necessary for the test. Check that with your solution, it would not be possible to do something like that, because it would try to find a contact in the database anyway.

Reply:
I removed all FK references in the defaults of the document fabricator